pkg/apis/storage/types.go

```diff
@@ -363,11 +363,14 @@ type CSIDriverSpec struct {
 	// most one token is empty string. To receive a new token after expiry,
 	// RequiresRepublish can be used to trigger NodePublishVolume periodically.
 	//
-	// This is an alpha feature and only available when the
+	// This is a beta feature and only available when the
 	// CSIServiceAccountToken feature is enabled.
 	//
 	// +optional
-	// +listType=atomic
+	// +patchMergeKey=audience
+	// +patchStrategy=merge
+	// +listMapKey=audience
+	// +listType=map
 	TokenRequests []TokenRequest
```

Comment:
do you have a reference for why this change is being made? do we expect multiple actors to coordinate on this field or to specify the entire list at once?

Reply:
in alpha, we didn't use `patchStrategy` because `patchMergeKey` doesn't support empty string in 1.20. talked to @apelisse, that is supported in 1.21 now.

yea, there could be multiple actors on this field.

Reply:
The issue is that patchStrategy/patchMergeKey gets baked into clients and affects the patch kubectl sends when invoking `kubectl apply`. Changing that here means a 1.20 kubectl will send the wrong content when updating this field via apply.

Reply:
Doesn't kubectl uses the openapi at that point to make the merge decision?

Reply:
can you verify that is true (e.g. use a 1.20 kubectl to client-side apply a change to a server running with this PR and see if it honors the new patch strategy)

Reply:
```
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:12:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.0-alpha.3.908+d2922316ed8131", GitCommit:"d2922316ed81319336024e9368f3ffb6ce6d981c", GitTreeState:"clean", BuildDate:"2021-03-04T18:07:21Z", GoVersion:"go1.16", Compiler:"gc", Platform:"linux/amd64"}
```

the following is the log when i am trying to apply a CSIDriver object(replaced audience "gcp" with "newgcp"). the new patch strategy is used and the calculation is correct. however, it failed to do so due to "immutable field" while the `TokenRequests` field is mutable.
```
I0304 10:33:17.444489 3247191 round_trippers.go:425] curl -k -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Authorization: Bearer <masked>" 'https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com'
I0304 10:33:17.455476 3247191 round_trippers.go:445] GET https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com 200 OK in 10 milliseconds
I0304 10:33:17.455516 3247191 round_trippers.go:451] Response Headers:
I0304 10:33:17.455531 3247191 round_trippers.go:454]     Date: Thu, 04 Mar 2021 18:33:17 GMT
I0304 10:33:17.455544 3247191 round_trippers.go:454]     Audit-Id: d88cdc8b-5985-450e-997c-e01a5cf0afb0
I0304 10:33:17.455554 3247191 round_trippers.go:454]     Cache-Control: no-cache, private
I0304 10:33:17.455565 3247191 round_trippers.go:454]     Content-Type: application/json
I0304 10:33:17.455578 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Flowschema-Uid: 66e63b9b-77dd-49b9-94f4-660891cf396c
I0304 10:33:17.455589 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Prioritylevel-Uid: 90666279-523f-47fc-9a23-dc12a60a3632
I0304 10:33:17.455617 3247191 round_trippers.go:454]     Content-Length: 1463
I0304 10:33:17.455695 3247191 request.go:1107] Response Body: {"kind":"CSIDriver","apiVersion":"storage.k8s.io/v1","metadata":{"name":"test.example.com","uid":"1c69bda6-18ea-4ed9-b552-acb56dc57d28","resourceVersion":"1703","creationTimestamp":"2021-03-04T18:31:08Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"gcp\"},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"},"managedFields":[{"manager":"kubectl-client-side-apply","operation":"Update","apiVersion":"storage.k8s.io/v1","time":"2021-03-04T18:31:08Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:kubectl.kubernetes.io/last-applied-configuration":{}}},"f:spec":{"f:attachRequired":{},"f:fsGroupPolicy":{},"f:podInfoOnMount":{},"f:requiresRepublish":{},"f:tokenRequests":{".":{},"k:{\"audience\":\"\"}":{".":{},"f:audience":{},"f:expirationSeconds":{}},"k:{\"audience\":\"gcp\"}":{".":{},"f:audience":{}}},"f:volumeLifecycleModes":{".":{},"v:\"Ephemeral\"":{},"v:\"Persistent\"":{}}}}}]},"spec":{"attachRequired":true,"podInfoOnMount":true,"volumeLifecycleModes":["Persistent","Ephemeral"],"fsGroupPolicy":"ReadWriteOnceWithFSType","tokenRequests":[{"audience":"gcp"},{"audience":"","expirationSeconds":3600}],"requiresRepublish":false}}
I0304 10:33:17.457010 3247191 request.go:1107] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"newgcp\"},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"}},"spec":{"$setElementOrder/tokenRequests":[{"audience":"newgcp"},{"audience":""}],"tokenRequests":[{"audience":"newgcp"},{"$patch":"delete","audience":"gcp"}]}}
I0304 10:33:17.457147 3247191 round_trippers.go:425] curl -k -v -XPATCH  -H "Accept: application/json" -H "Content-Type: application/strategic-merge-patch+json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Authorization: Bearer <masked>" 'https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply'
I0304 10:33:17.472575 3247191 round_trippers.go:445] PATCH https://34.83.244.149/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply 422 Unprocessable Entity in 15 milliseconds
I0304 10:33:17.472606 3247191 round_trippers.go:451] Response Headers:
I0304 10:33:17.472622 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Flowschema-Uid: 66e63b9b-77dd-49b9-94f4-660891cf396c
I0304 10:33:17.472639 3247191 round_trippers.go:454]     X-Kubernetes-Pf-Prioritylevel-Uid: 90666279-523f-47fc-9a23-dc12a60a3632
I0304 10:33:17.472655 3247191 round_trippers.go:454]     Content-Length: 1394
I0304 10:33:17.472670 3247191 round_trippers.go:454]     Date: Thu, 04 Mar 2021 18:33:17 GMT
I0304 10:33:17.472686 3247191 round_trippers.go:454]     Audit-Id: bed10d8a-fb41-446e-ad32-33edcd760dce
I0304 10:33:17.472701 3247191 round_trippers.go:454]     Cache-Control: no-cache, private
I0304 10:33:17.472717 3247191 round_trippers.go:454]     Content-Type: application/json
I0304 10:33:17.473462 3247191 request.go:1107] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"CSIDriver.storage.k8s.io \"test.example.com\" is invalid: spec: Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{\"Persistent\", \"Ephemeral\"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:\"newgcp\", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:\"\", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable","reason":"Invalid","details":{"name":"test.example.com","group":"storage.k8s.io","kind":"CSIDriver","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{\"Persistent\", \"Ephemeral\"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:\"newgcp\", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:\"\", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable","field":"spec"}]},"code":422}
F0304 10:33:17.474004 3247191 helpers.go:115] The CSIDriver "test.example.com" is invalid: spec: Invalid value: storage.CSIDriverSpec{AttachRequired:(*bool)(0xc0016e36c8), FSGroupPolicy:(*storage.FSGroupPolicy)(0xc013a42690), PodInfoOnMount:(*bool)(0xc0016e36c9), VolumeLifecycleModes:[]storage.VolumeLifecycleMode{"Persistent", "Ephemeral"}, StorageCapacity:(*bool)(nil), TokenRequests:[]storage.TokenRequest{storage.TokenRequest{Audience:"newgcp", ExpirationSeconds:(*int64)(nil)}, storage.TokenRequest{Audience:"", ExpirationSeconds:(*int64)(0xc0016e3700)}}, RequiresRepublish:(*bool)(0xc0016e3708)}: field is immutable
```

Reply:
`CSIDriverSpec` was immutable. changing two fields added by this feature gate to mutable. it is safe to make these two mutable because every `NodePublishVolume` call would read the value. cc @msau42 

the patch strategy is working according to the above log. would test it again with the new changes

Reply:
```
I0304 16:53:27.994030 3587476 request.go:1107] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"CSIDriver\",\"metadata\":{\"annotations\":{},\"name\":\"test.example.com\"},\"spec\":{\"attachRequired\":true,\"podInfoOnMount\":true,\"tokenRequests\":[{\"audience\":\"gcp\",\"expirationSeconds\":7200},{\"audience\":\"\",\"expirationSeconds\":3600}],\"volumeLifecycleModes\":[\"Persistent\",\"Ephemeral\"]}}\n"}},"spec":{"$setElementOrder/tokenRequests":[{"audience":"gcp"},{"audience":""}],"tokenRequests":[{"audience":"gcp","expirationSeconds":7200}]}}
I0304 16:53:27.994123 3587476 round_trippers.go:425] curl -k -v -XPATCH  -H "Accept: application/json" -H "User-Agent: kubectl/v1.20.4 (linux/amd64) kubernetes/e87da0b" -H "Content-Type: application/strategic-merge-patch+json" -H "Authorization: Bearer <masked>" 'https://34.82.67.178/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply'
I0304 16:53:28.010485 3587476 round_trippers.go:445] PATCH https://34.82.67.178/apis/storage.k8s.io/v1/csidrivers/test.example.com?fieldManager=kubectl-client-side-apply 200 OK in 16 milliseconds
```

looks working :)